File: Java/src/stacks_and_queues/Queue.java
package stacks_and_queues;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Chris Keyser on 3/17/2016.
 * Simple queue implementation using generics and a list to represent the data
 * Handles I/O in FIFO order because it's a queue
 */
public class Queue<T> {
    private List<T> data;

    public Queue() {
        this.data = new ArrayList<>();
    }

    /**
     * initializes Queue to have information in the order of given List
     */
    public Queue(List<T> data) {
        this.data = new ArrayList<>(data);
    }

    /**
     * pushes object onto back of queue
     */
    public void enqueue(T obj) {
        this.data.add(obj);
    }

    /**
     * pops an object off the front of the queue
     */
    public T dequeue() {
        return !isEmpty() ? this.data.remove(0) : null;
    }

    /**
     * returns number of objects in queue
     */
    public int size() {
        return this.data.size();
    }

    /**
     * returns a boolean representing whether the queue is empty
     */
    public boolean isEmpty() {
        return this.data.isEmpty();
    }

    /**
     * determines whether the queue contains object obj
     */
    public boolean contains(T obj) {
        return data.contains(obj);
    }
}

File: haskell/src/Stack/Queue.hs
module Queue
( Queue
, newQueue
, enqueue
, dequeue
) where

type Queue a = [a]

-- Initialize a queue with the given value
newQueue :: a -> Queue a
newQueue x = [x]

-- enqueue an item into the back of the queue
enqueue :: Queue a -> a -> Queue a
enqueue s x  = s ++ [x]

-- dequeue an item from the front of the queue
-- returns both the dequeued item and the new state of the queue
dequeue :: Queue a -> (a, Queue a)
dequeue s = (head s, tail s)
